### ISSUE TITLE
MDC lost when processes are resumed on a different thread after hitting a suspension point

### ISSUE THREAD

**Issue by 22dasi1bwi** [micronaut-projects/micronaut-core#6113]
### Expected Behavior

Under heavy load MDC keeps correct information after first suspension point.

### Actual Behaviour

Under heavy load MDC loses correct information after first suspension point. See [example application](https://github.com/22dasi1bwi/micronaut-coroutines-context-loss-on-thread-switch) for more information.

Is there a preferred way on how to correctly do this with Reactor in Micronaut 3.0?

### Steps To Reproduce

All required steps to reproduce the behavior are provided in the [example application](https://github.com/22dasi1bwi/micronaut-coroutines-context-loss-on-thread-switch) . Also the background information are described there.

See [related issue](https://github.com/micronaut-projects/micronaut-core/issues/5656) for even more details.

### Environment Information

Operating System: Arch Linux
JDK Version: 11.0.12

### Example Application

https://github.com/22dasi1bwi/micronaut-coroutines-context-loss-on-thread-switch

### Version

3.0.0

**Comment by dstepanov** [comment]
Can you please explain:
```
After upgrading to Micronaut 3.0 the two classes required to get things working in Micronaut 2.5.x com.example.ServerRequestContextFilterCopy.kt and com.example.ServerRequestContextInstrumenterCopy.kt are preventing the MDC to have proper information once processes are resumed on a different thread after hitting a suspension point.

The removal of those two classes makes stress testing (see Howto and execute_benchmark.sh) only work at a rate of approximately 40%.
```

Are you saying that it does work with those classes added but without it works only 40%?

**Comment by 22dasi1bwi** [comment]
It worked with those two classes in Micronaut version 2.x (100%) . But it has an error rate of 100% if I keep those two classes in Micronaut version 3.x . 
In Micronaut version 3.x I had to remove those two classes to get context propagation work at all. With those two classes gone the success rate raised from 0% to like 40%.

**Comment by dstepanov** [comment]
This is fixed now in 3.1 and BTW `client.exchange(request, String::class.java).awaitFirst()` can be non-blocking `client.exchange(request, String::class.java).asFlow().single()`

**Comment by 22dasi1bwi** [comment]
@dstepanov 
Thanks for the fix. Can't wait to try it out! :smiley: 

I thought `awaitFirst()` would be non-blocking, too? That's at least what the documentation says:

`Awaits the first value from the given publisher without blocking the thread and returns the resulting value, or, if the publisher has produced an error, throws the corresponding exception.`

Would you mind explaining?

Thanks!

**Comment by dstepanov** [comment]
Ah now I look at it and it doesn't look blocking 🤷‍♂️ Sorry
You can try the 3.1 snapshot

**Comment by 22dasi1bwi** [comment]
Alright. I'm glad we're on the same page :smiley: 

I tried out 3.1 with my sample project and it is indeed working as expected! Nice :) . I updated the [readme](https://github.com/22dasi1bwi/micronaut-coroutines-context-loss-on-thread-switch/blob/master/README.md) based on the newest changes, just in case somebody else is stumbling over this issue.

Thank you very much for the support on this @dstepanov . It's highly appreciated.

Cheers!
